frame/merkle-mountain-range/primitives/src/lib.rs

```diff
@@ -274,15 +279,114 @@ pub struct Proof<Hash> {
 	pub items: Vec<Hash>,
 }
 
+/// Merkle Mountain Range operation error.
+#[derive(RuntimeDebug, codec::Encode, codec::Decode, PartialEq, Eq)]
+pub enum Error {
+	/// Error while pushing new node.
+	Push,
+	/// Error getting the new root.
+	GetRoot,
+	/// Error commiting changes.
+	Commit,
+	/// Error during proof generation.
+	GenerateProof,
+	/// Proof verification error.
+	Verify,
+	/// Leaf not found in the storage.
+	LeafNotFound,
+}
+
+impl Error {
+	#![allow(unused_variables)]
+	/// Consume given error `e` with `self` and generate a native log entry with error details.
+	pub fn log_error(self, e: impl fmt::Debug) -> Self {
+		debug::native::error!("[{:?}] MMR error: {:?}", self, e);
+		self
+	}
+
+	/// Consume given error `e` with `self` and generate a native log entry with error details.
+	pub fn log_debug(self, e: impl fmt::Debug) -> Self {
+		debug::native::debug!("[{:?}] MMR error: {:?}", self, e);
+		self
+	}
+}
+
+/// A helper type to allow using arbitrary SCALE-encoded leaf data in the RuntimeApi.
+///
+/// The point is to be able to verify MMR proofs from external MMRs, where we don't
+/// know the exact leaf type, but it's enough for us to have it SCALE-encoded.
+///
+/// Note the leaf type should be encoded in it's compact form when passed through this type.
+/// See [FullLeaf] documentation for details.
+///
+/// This type is SCALE-compatible with `Vec<u8>` encoding. I.e. you must encode your leaf twice:
+/// ```rust,ignore
+/// let encoded: Vec<u8> = my_leaf.encode();
+/// let opaque: Vec<u8> = encoded.encode();
+/// let decoded = OpaqueLeaf::decode(&mut &*opaque);
+/// ```
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+#[derive(RuntimeDebug, Clone, PartialEq, codec::Decode)]
+pub struct OpaqueLeaf(
+	/// Leaf type encoded in it's compact form.
+	#[cfg_attr(feature = "std", serde(with = "sp_core::bytes"))]
+	pub Vec<u8>
+);
+
+impl OpaqueLeaf {
+	/// Convert a concrete MMR leaf into an opaque type.
+	pub fn from_leaf<T: FullLeaf>(leaf: T) -> Self {
+		let encoded_leaf = leaf.using_encoded(|d| d.to_vec(), true);
+		// OpaqueLeaf must be SCALE-compatible with `Vec<u8>`.
+		// Simply using raw encoded bytes don't work, cause we don't know the
+		// length of the expected data.
+		let encoded_vec = codec::Encode::encode(&encoded_leaf);
+		OpaqueLeaf(encoded_vec)
+	}
+}
+
+impl FullLeaf for OpaqueLeaf {
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F, _compact: bool) -> R {
+		f(&self.0)
```

Comment:
this type encoding/decoding is bit confusing to me.
The encoding is identity but the decoding is not.
e.g. `encode(decode(encode(OpaqueLeaf(something))) != encode(OpaqueLeaf(something))`

I think it would make more sense to have a manual implementation of Decode.
Or to change to something like this:
```suggestion
	pub fn from_leaf<T: FullLeaf>(leaf: T) -> Self {
		let encoded_leaf = leaf.using_encoded(|d| d.to_vec(), true);
		OpaqueLeaf(encoded_leaf)
	}
}

impl FullLeaf for OpaqueLeaf {
	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F, _compact: bool) -> R {
		f(&self.0.encode())
```

Reply:
I was writing an elaborate answer why this is correct, just to learn that it's all wrong :) Thanks for being persistent with this.

So my initialy motivation was to have `OpaqueLeaf` be `Vec<u8>` SCALE-compatible, but pretend to be the "unwraped" (no length prefix) encoding for MMR struct. I've realised that this is not really needed, and I can relax the `FullLeaf` requirement and have `OpaqueLeaf` be completely independent from SCALE.

Please see the updated version now. The runtime API just accepts a raw `Vec<u8>`, and we just wrap this into `OpaqueLeaf` to get custom `FullLeaf` implementation (the default one (blanket impl for `codec::Encode`) would re-encode the vec, which is what I wanted to avoid).